### PR TITLE
arm64: dts: qcom: pmi8950: enable vibrator haptics

### PR DESCRIPTION
**Pull request by M0Rf30**
from dmesg:
```
[   14.312943] input: spmi_haptics as /devices/platform/soc@0/200f000.spmi/spmi-0/0-03/200f000.spmi:pmic@3:haptics@c000/input/input4
```
with fftest
```
xiaomi-vince:~$ sudo fftest /dev/input/event4
Force feedback test program.
HOLD FIRMLY YOUR WHEEL OR JOYSTICK TO PREVENT DAMAGES

Device /dev/input/event4 opened
Features:
  * Absolute axes:
    [00 00 00 00 00 00 00 00 ]
  * Relative axes:
    [00 00 ]
  * Force feedback effects types: Periodic, Rumble, Gain,
    Force feedback periodic effects: Square, Triangle, Sine,
    [00 00 00 00 00 00 00 00 00 00 03 07 01 00 00 00 ]
  * Number of simultaneous effects: 16

Setting master gain to 75% ... OK
Uploading effect #0 (Periodic sinusoidal) ... OK (id 0)
Uploading effect #1 (Constant) ... Error: Invalid argument
Uploading effect #2 (Spring) ... Error: Invalid argument
Uploading effect #3 (Damper) ... Error: Invalid argument
Uploading effect #4 (Strong rumble, with heavy motor) ... OK (id 1)
Uploading effect #5 (Weak rumble, with light motor) ... OK (id 2)
```

### CODE REVIEW

File: arch/arm64/boot/dts/qcom/pmi8950.dtsi [hunk 1]
```diff
@@ -199,6 +200,17 @@
...
 
+		pmi8950_haptics: haptics@c000 {
+			compatible = "qcom,pmi8950-haptics", "qcom,spmi-haptics";
+			reg = <0xc000>;
+
+			interrupts = <0x3 0xc0 0x0 IRQ_TYPE_EDGE_BOTH>,
+				     <0x3 0xc0 0x1 IRQ_TYPE_EDGE_BOTH>;
+			interrupt-names = "short", "play";
+
+			qcom,play-mode = <HAP_PLAY_BUFFER>;
```
Comment (barni2000) [inline arch/arm64/boot/dts/qcom/pmi8950.dtsi:211]:
```suggestion
			qcom,play-mode = <HAP_PLAY_BUFFER>;
			
			status = "disabled";
```

**Comment by barni2000** [comment]
The previous attempt had weak rumble is it working fine now?

**Comment by M0Rf30** [comment]
> The previous attempt had weak rumble. Is it working fine now?

Same weak rumble, I only synced with 6.11 and cleaned up.

**Comment by alikates** [comment]
Hi, i had some initial work on adding support for Eccentric Rotating Mass (cheap motor + counterweight) in  https://github.com/msm8953-mainline/linux/commit/3088192deb7f6a4ece90ebd42329f53876512fbb. Idk if you already have this support in place, but otherwide it may be relevant for you 🙂

File: arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts [hunk 2]
```diff
@@ -206,6 +206,15 @@
 	regulator-min-microvolt = <1200000>;
 };
 
+&pmi8950_haptics {
+	qcom,actuator-type = <HAP_TYPE_LRA>;
+	qcom,brake-pattern = [0x3 0x3 0x0 0x0];
+	qcom,wave-play-rate-us = <5263>;
+	qcom,wave-shape = <HAP_WAVE_SINE>;
```
Comment (barni2000) [inline arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts:213]:
```suggestion
	qcom,wave-shape = <HAP_WAVE_SQUARE>;
```

File: arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts [hunk 3]
```diff
@@ -206,6 +206,15 @@
 	regulator-min-microvolt = <1200000>;
 };
 
+&pmi8950_haptics {
+	qcom,actuator-type = <HAP_TYPE_LRA>;
```
Comment (barni2000) [inline arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts:210]:
```suggestion
	qcom,actuator-type = <HAP_TYPE_ERM>;
```

File: arch/arm64/boot/dts/qcom/pmi8950.dtsi [hunk 4]
```diff
@@ -199,6 +200,18 @@
...
 		};
 
+		pmi8950_haptics: haptics@c000 {
+			compatible = "qcom,pmi8950-haptics", "qcom,spmi-haptics";
+			reg = <0xc000>;
+
+			interrupts = <0x3 0xc0 0x0 IRQ_TYPE_EDGE_BOTH>,
+				     <0x3 0xc0 0x1 IRQ_TYPE_EDGE_BOTH>;
+			interrupt-names = "short", "play";
+			qcom,play-mode = <HAP_PLAY_BUFFER>;
```
Comment (barni2000) [inline arch/arm64/boot/dts/qcom/pmi8950.dtsi:210]:
```suggestion
			qcom,play-mode = <HAP_PLAY_DIRECT>;
```

File: arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts [hunk 5]
```diff
@@ -206,6 +206,15 @@
...
+	qcom,brake-pattern = [0x3 0x3 0x0 0x0];
```
Comment (barni2000) [inline arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts:211]:
```suggestion
	qcom,brake-pattern = <0x3 0x3 0x0 0x0>;
```

**Review by barni2000 — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/pmi8950.dtsi (hunk 4 above)
Comment (barni2000) [reply arch/arm64/boot/dts/qcom/pmi8950.dtsi:210]: It seems driver is not supports direct mode yet.

File: arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts (hunk 3 above)
Comment (barni2000) [reply arch/arm64/boot/dts/qcom/msm8953-xiaomi-vince.dts:210]: ERM is not supported by the driver

**Comment by barni2000** [comment]
Driver should be updated a bit to support ERM and DIRECT mode but it will work.

**Pull request closed without merging**